clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1001,311 +1002,580 @@ static ParamDesc makeParamDesc(const FieldDecl *Src, QualType Ty) {
-// Returns true if all arguments are successfully built.
-static bool buildArgTys(ASTContext &Context, CXXRecordDecl *KernelObj,
-                        SmallVectorImpl<ParamDesc> &ParamDescs) {
-  auto CreateAndAddPrmDsc = [&](const FieldDecl *Fld, const QualType &ArgType) {
-    // Create a parameter descriptor and append it to the result
-    ParamDescs.push_back(makeParamDesc(Fld, ArgType));
-  };
-
-  // Creates a parameter descriptor for SYCL special object - SYCL accessor or
-  // sampler.
+// The first template argument to the kernel function is used to identify the
+// kernel itself.
+static QualType calculateKernelNameType(ASTContext &Ctx,
+                                        FunctionDecl *KernelCallerFunc) {
+  // TODO: Not sure what the 'fully qualified type's purpose is here, the type
+  // itself should have its full qualified name, so figure out what the purpose
+  // is.
+  const TemplateArgumentList *TAL =
+      KernelCallerFunc->getTemplateSpecializationArgs();
+  return TypeName::getFullyQualifiedType(TAL->get(0).getAsType(), Ctx,
+                                         /*WithGlobalNSPrefix=*/true);
+}
+
+// Gets a name for the kernel caller func, calculated from the first template
+// argument.
+static std::string constructKernelName(Sema &S, FunctionDecl *KernelCallerFunc,
+                                       MangleContext &MC, bool StableName) {
+  QualType KernelNameType =
+      calculateKernelNameType(S.getASTContext(), KernelCallerFunc);
+
+  if (StableName)
+    return PredefinedExpr::ComputeName(S.getASTContext(),
+                                       PredefinedExpr::UniqueStableNameType,
+                                       KernelNameType);
+  SmallString<256> Result;
+  llvm::raw_svector_ostream Out(Result);
+
+  MC.mangleTypeName(KernelNameType, Out);
+  return std::string(Out.str());
+}
+
+// anonymous namespace so these don't get linkage.
+namespace {
+
+QualType getItemType(const FieldDecl *FD) { return FD->getType(); }
+QualType getItemType(const CXXBaseSpecifier &BS) { return BS.getType(); }
+
+// Implements the 'for-each-visitor'  pattern.
+template <typename ParentTy, typename... Handlers>
+static void VisitAccessorWrapper(CXXRecordDecl *Owner, ParentTy &Parent,
+                                 CXXRecordDecl *Wrapper,
+                                 Handlers &... handlers);
+
+template <typename RangeTy, typename... Handlers>
+static void VisitAccessorWrapperHelper(CXXRecordDecl *Owner, RangeTy Range,
+                                       Handlers &... handlers) {
+  for (const auto &Item : Range) {
+    QualType ItemTy = getItemType(Item);
+    if (Util::isSyclAccessorType(ItemTy))
+      (void)std::initializer_list<int>{
+          (handlers.handleSyclAccessorType(Item, ItemTy), 0)...};
+    else if (Util::isSyclStreamType(ItemTy))
+      (void)std::initializer_list<int>{
+          (handlers.handleSyclStreamType(Item, ItemTy), 0)...};
+    else if (ItemTy->isStructureOrClassType()) {
```

Comment:
Hmm... I didn't see in the old code it recursing through arrays, are we intending to split them out as well?

In that case, we probably need to better handle the offsets.

Reply:
Issue captured here: https://github.com/intel/llvm/issues/1567

Reply:
It is quite natural to recurse through structs to any level to discover accessors and the new visitor model does that.
The existing implementation does not deal with arrays at all so you won't see any code there.
My support for arrays accounted for arrays of accessors, i.e., the array type is one of the special sycl types. When such an array type was encountered it would "unroll" each element of the array and generate a separate descriptor for it. It did not try to handle arrays of other struct types that contained accessors (or arrays of accessors) as members, or more deeply nested members.
I suggest we support arrays of accessors only, not arrays of structs that may in turn contain accessors.
If so, the proposed addition to the visitor function will do the job.

Reply:
I think we'd have to support both, arrays of structs of accessors, or even arrays of structs of arrays of accessors.

It is a presumably fairly easy addition however, but I'll hold it off for another patch, we have a fairly sizable change coming from  @elizabethandrews  on how parameters are passed, so I'd like to hold off on arrays until after that.